File: app/src/main/java/br/com/stonesdk/sdkdemo/activities/TransactionListActivity.java
package br.com.stonesdk.sdkdemo.activities;

import android.app.AlertDialog;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

import br.com.stone.sdk.android.error.StoneStatus;
import br.com.stone.sdk.core.providers.interfaces.StoneCallbackInterface;
import br.com.stone.sdk.hardware.providers.PosPrintProvider;
import br.com.stone.sdk.payment.database.dao.transaction.TransactionDAO;
import br.com.stone.sdk.payment.database.models.transaction.TransactionObject;
import br.com.stone.sdk.payment.enums.ReceiptType;
import br.com.stone.sdk.payment.providers.CancellationProvider;
import br.com.stone.sdk.payment.providers.CaptureTransactionProvider;
import br.com.stone.sdk.payment.providers.PosPrintReceiptProvider;
import br.com.stone.sdk.payment.providers.SendEmailTransactionProvider;
import br.com.stone.sdk.payment.providers.model.email.Contact;
import br.com.stonesdk.sdkdemo.R;

public class TransactionListActivity extends AppCompatActivity implements OnItemClickListener {

    ListView listView;
    List<TransactionObject> transactionObjects;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_transaction_list);

        listView = findViewById(R.id.listTransactionActivity);

        TransactionDAO transactionDAO = new TransactionDAO(getApplicationContext());
        transactionObjects = transactionDAO.getAllTransactionsOrderByIdDesc();
        String[] rowOfList = new String[transactionObjects.size()];
        for (int i = 0; i < transactionObjects.size(); i++) {
            rowOfList[i] = String.format("%s=%s\n%s", transactionObjects.get(i).getIdFromBase(), transactionObjects.get(i).getAmount(), transactionObjects.get(i).getTransactionStatus());
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, android.R.id.text1, rowOfList);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(this);
    }

    public void onItemClick(AdapterView<?> parent, View view, final int position, long id) {
        final TransactionObject selectedTransaction = transactionObjects.get(position);
        ArrayList<String> optionsList = new ArrayList<String>() {{
            add("[POS] Imprimir via do estabelecimento");
            add("[POS] Imprimir via do cliente");
            add("[POS] Imprimir comprovante customizado");
            add("Cancelar");
            add("Enviar e-mail via do cliente");
            add("Enviar e-mail via do estabelecimento");
        }};
        if (!selectedTransaction.isCapture()) {
            optionsList.add("Capturar Transação");
        }
        String[] options = new String[optionsList.size()];
        optionsList.toArray(options);

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.list_dialog_title)
                .setItems(options, (dialog, which) -> {
                    switch (which) {
                        case 0:
                            printReceipt(ReceiptType.MERCHANT, selectedTransaction);
                            break;

                        case 2:
                            printReceipt(ReceiptType.CLIENT, selectedTransaction);
                            break;

                        case 3:
                            final PosPrintProvider customPosPrintProvider = new PosPrintProvider(TransactionListActivity.this);
                            customPosPrintProvider.addLine("");
                            customPosPrintProvider.addLine("PAN : " + selectedTransaction.getCardHolderNumber());
                            customPosPrintProvider.addLine("DATE/TIME : " + selectedTransaction.getDate() + " " + selectedTransaction.getTime());
                            customPosPrintProvider.addLine("AMOUNT : " + selectedTransaction.getAmount());
                            customPosPrintProvider.addLine("ATK : " + selectedTransaction.getAcquirerTransactionKey());
                            customPosPrintProvider.addLine("");
                            customPosPrintProvider.addBitmap(BitmapFactory.decodeResource(getResources(), R.drawable.signature));
                            customPosPrintProvider.print(new StoneCallbackInterface() {
                                @Override
                                public void onSuccess() {
                                    Toast.makeText(TransactionListActivity.this, "Recibo impresso", Toast.LENGTH_SHORT).show();
                                }

                                @Override
                                public void onError(@Nullable StoneStatus stoneStatus) {
                                    assert stoneStatus != null;
                                    Toast.makeText(TransactionListActivity.this, "Erro ao imprimir: " + stoneStatus.getMessage(), Toast.LENGTH_SHORT).show();
                                }
                            });
                            break;

                        case 4:
                            final CancellationProvider cancellationProvider = new CancellationProvider(TransactionListActivity.this, selectedTransaction);
                            cancellationProvider.setConnectionCallback(new StoneCallbackInterface() {
                                public void onSuccess() {
                                    Toast.makeText(getApplicationContext(),
                                            cancellationProvider.getMessageFromAuthorize(),
                                            Toast.LENGTH_SHORT).show();
                                    finish();
                                }

                                public void onError(@Nullable StoneStatus stoneStatus) {
                                    Toast.makeText(getApplicationContext(), "Um erro ocorreu durante o cancelamento com a transacao de id: " + selectedTransaction.getIdFromBase(), Toast.LENGTH_SHORT).show();
                                }
                            });
                            cancellationProvider.execute();
                            break;
                        case 5:
                            sendReceipt(selectedTransaction, ReceiptType.CLIENT);
                            break;
                        case 6:
                            sendReceipt(selectedTransaction, ReceiptType.MERCHANT);
                            break;
                        case 7:
                            final CaptureTransactionProvider provider = new CaptureTransactionProvider(TransactionListActivity.this, selectedTransaction);
                            provider.setConnectionCallback(new StoneCallbackInterface() {
                                @Override
                                public void onSuccess() {
                                    Toast.makeText(TransactionListActivity.this, "Transação Capturada com sucesso!",
                                            Toast.LENGTH_SHORT).show();
                                }

                                @Override
                                public void onError(@Nullable StoneStatus stoneStatus) {
                                    assert stoneStatus != null;
                                    Toast.makeText(TransactionListActivity.this, "Ocorreu um erro captura da transacao: " +
                                                    stoneStatus.getMessage(),
                                            Toast.LENGTH_SHORT).show();

                                }
                            });

                            provider.execute();
                            break;
                    }
                });
        builder.create();
        builder.show();
    }

    private void sendReceipt(TransactionObject selectedTransaction, ReceiptType receiptType) {
        SendEmailTransactionProvider sendEmailProvider = new SendEmailTransactionProvider(TransactionListActivity.this, selectedTransaction);
        sendEmailProvider.setReceiptType(receiptType);
        sendEmailProvider.addTo(new Contact("cliente@gmail.com", "Nome do Cliente"));
        sendEmailProvider.setFrom(new Contact("loja@gmail.com", "Nome do Estabelecimento"));
        sendEmailProvider.setConnectionCallback(new StoneCallbackInterface() {
            public void onSuccess() {
                Toast.makeText(getApplicationContext(), "Enviado com sucesso", Toast.LENGTH_LONG).show();
            }

            public void onError(@Nullable StoneStatus stoneStatus) {
                Toast.makeText(getApplicationContext(), "Nao enviado", Toast.LENGTH_LONG).show();
            }
        });
        sendEmailProvider.execute();
    }

    private void printReceipt(ReceiptType receiptType, TransactionObject transactionObject) {
        PosPrintReceiptProvider posPrintReceiptProvider = new PosPrintReceiptProvider(this, transactionObject, receiptType);
        posPrintReceiptProvider.print(new StoneCallbackInterface() {
            @Override
            public void onSuccess() {
                Toast.makeText(getApplicationContext(), "Recibo impresso", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onError(@Nullable StoneStatus stoneStatus) {
                assert stoneStatus != null;
                Toast.makeText(getApplicationContext(), "Erro ao imprimir: " + stoneStatus.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/br/com/stonesdk/sdkdemo/activities/TransactionListActivity.kt
package br.com.stonesdk.sdkdemo.activities

import android.app.AlertDialog
import android.content.DialogInterface
import android.graphics.BitmapFactory
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import br.com.stone.posandroid.providers.PosPrintProvider
import br.com.stone.posandroid.providers.PosPrintReceiptProvider
import br.com.stonesdk.sdkdemo.R
import br.com.stonesdk.sdkdemo.controller.PrintController
import br.com.stonesdk.sdkdemo.databinding.ActivityTransactionBinding
import br.com.stonesdk.sdkdemo.databinding.ActivityTransactionListBinding
import stone.application.enums.ReceiptType
import stone.application.interfaces.StoneCallbackInterface
import stone.database.transaction.TransactionDAO
import stone.database.transaction.TransactionObject
import stone.providers.CancellationProvider
import stone.providers.CaptureTransactionProvider
import stone.providers.PrintProvider
import stone.providers.SendEmailTransactionProvider
import stone.repository.remote.email.pombo.email.Contact
import stone.utils.PrintObject
import stone.utils.Stone

class TransactionListActivity() : AppCompatActivity(), OnItemClickListener {

    private lateinit var binding: ActivityTransactionListBinding




    var listView: ListView? = null
    var transactionObjects: List<TransactionObject>? = null

    override fun onCreate(savedInstanceState: Bundle?) {




        super.onCreate(savedInstanceState)
        binding = ActivityTransactionListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        listView = binding.listTransactionActivity

        // acessa todas as transacoes do banco de dados
        val transactionDAO = TransactionDAO(applicationContext)
        // cria uma lista com todas as transacoes
        transactionObjects = transactionDAO.getAllTransactionsOrderByIdDesc()

        // exibe todas as transações (neste caso valor e status) para o usuario
        val rowOfList = arrayOfNulls<String>(
            transactionObjects!!.size
        )
        for (i in transactionObjects!!.indices) {
            rowOfList[i] = String.format(
                "%s=%s\n%s",
                transactionObjects!![i].idFromBase,
                transactionObjects!![i].amount,
                transactionObjects!![i].transactionStatus
            )
        }
        val adapter =
            ArrayAdapter(this, android.R.layout.simple_list_item_1, android.R.id.text1, rowOfList)
        listView!!.setAdapter(adapter)
        listView!!.setOnItemClickListener(this)
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
        val selectedTransaction = transactionObjects!![position]
        val optionsList: ArrayList<String?> = object : ArrayList<String?>() {
            init {
                add("[Pinpad] Imprimir comprovante")
                add("[POS] Imprimir via do estabelecimento")
                add("[POS] Imprimir via do cliente")
                add("[POS] Imprimir comprovante customizado")
                add("Cancelar")
                add("Enviar via do cliente")
                add("Enviar via do estabelecimento")
            }
        }
        if (!selectedTransaction.isCapture) {
            optionsList.add("Capturar Transação")
        }
        val options = arrayOfNulls<String>(optionsList.size)
        optionsList.toArray(options)

        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.list_dialog_title)
            .setItems(options, DialogInterface.OnClickListener { dialog, which ->
                when (which) {
                    0 -> try {
                        // lógica da impressão
                        val listToPrint: MutableList<PrintObject> = ArrayList()
                        var i = 0
                        while (i < 10) {
                            listToPrint.add(
                                PrintObject(
                                    "Teste de impressão linha $i",
                                    PrintObject.MEDIUM,
                                    PrintObject.CENTER
                                )
                            )
                            i++
                        }
                        // Stone.getPinpadFromListAt(0) eh o pinpad conectado, que esta na posicao zero.
                        val printProvider = PrintProvider(
                            this@TransactionListActivity,
                            listToPrint,
                            Stone.getPinpadFromListAt(0)
                        )
                        printProvider.dialogMessage = "Imprimindo..."
                        printProvider.connectionCallback = object : StoneCallbackInterface {
                            override fun onSuccess() {
                                Toast.makeText(
                                    applicationContext,
                                    "Impressão realizada com sucesso",
                                    Toast.LENGTH_SHORT
                                ).show()
                                finish()
                            }

                            override fun onError() {
                                Toast.makeText(
                                    applicationContext,
                                    "Um erro ocorreu durante a impressão",
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                        printProvider.execute()
                    } catch (outException: IndexOutOfBoundsException) {
                        Toast.makeText(
                            applicationContext,
                            "Conecte-se a um pinpad.",
                            Toast.LENGTH_SHORT
                        ).show()
                    } catch (e: Exception) {
                        Toast.makeText(
                            applicationContext,
                            "Houve um erro inesperado. Tente novamente mais tarde.",
                            Toast.LENGTH_SHORT
                        ).show()
                        e.printStackTrace()
                    }

                    1 -> printReceipt(ReceiptType.MERCHANT, selectedTransaction)
                    2 -> printReceipt(ReceiptType.CLIENT, selectedTransaction)
                    3 -> {
                        // Impressão customizada
                        val customPosPrintProvider = PosPrintProvider(this@TransactionListActivity)
                        customPosPrintProvider.addLine("Stone")
                        customPosPrintProvider.addLine("Comprovante customizado")
                        customPosPrintProvider.addLine("")
                        customPosPrintProvider.addLine("PAN : " + selectedTransaction.cardHolderNumber)
                        customPosPrintProvider.addLine("DATE/TIME : " + selectedTransaction.date + " " + selectedTransaction.time)
                        customPosPrintProvider.addLine("AMOUNT : " + selectedTransaction.amount)
                        customPosPrintProvider.addLine("ATK : " + selectedTransaction.acquirerTransactionKey)
                        customPosPrintProvider.addLine("")
                        customPosPrintProvider.addLine("Signature")
                        customPosPrintProvider.addBitmap(
                            BitmapFactory.decodeResource(
                                resources,
                                R.drawable.signature
                            )
                        )
                        customPosPrintProvider.connectionCallback =
                            object : StoneCallbackInterface {
                                override fun onSuccess() {
                                    Toast.makeText(
                                        this@TransactionListActivity,
                                        "Recibo impresso",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }

                                override fun onError() {
                                    Toast.makeText(
                                        this@TransactionListActivity,
                                        "Erro ao imprimir: " + customPosPrintProvider.listOfErrors,
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }
                            }
                        customPosPrintProvider.execute()
                    }

                    4 -> {
                        val cancellationProvider =
                            CancellationProvider(this@TransactionListActivity, selectedTransaction)
                        cancellationProvider.dialogMessage = "Cancelando..."
                        cancellationProvider.connectionCallback = object : StoneCallbackInterface {
                            // chamada de retorno.
                            override fun onSuccess() {
                                Toast.makeText(
                                    applicationContext,
                                    cancellationProvider.messageFromAuthorize,
                                    Toast.LENGTH_SHORT
                                ).show()
                                finish()
                            }

                            override fun onError() {
                                Toast.makeText(
                                    applicationContext,
                                    "Um erro ocorreu durante o cancelamento com a transacao de id: " + selectedTransaction.idFromBase,
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                        cancellationProvider.execute()
                    }

                    5 -> sendReceipt(selectedTransaction, ReceiptType.CLIENT)
                    6 -> sendReceipt(selectedTransaction, ReceiptType.MERCHANT)
                    7 -> {
                        val provider = CaptureTransactionProvider(
                            this@TransactionListActivity,
                            selectedTransaction
                        )
                        provider.dialogMessage = "Efetuando Captura..."
                        provider.connectionCallback = object : StoneCallbackInterface {
                            override fun onSuccess() {
                                Toast.makeText(
                                    this@TransactionListActivity, "Transação " +
                                            "Capturada com sucesso!",
                                    Toast.LENGTH_SHORT
                                ).show()
                            }

                            override fun onError() {
                                Toast.makeText(
                                    this@TransactionListActivity, ("Ocorreu um " +
                                            "erro captura da transacao: " +
                                            provider.listOfErrors),
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }

                        provider.execute()
                    }
                }
            })
        builder.create()
        builder.show()
    }

    private fun sendReceipt(selectedTransaction: TransactionObject, receiptType: ReceiptType) {
        val sendEmailProvider =
            SendEmailTransactionProvider(this@TransactionListActivity, selectedTransaction)
        sendEmailProvider.receiptType = receiptType
        sendEmailProvider.addTo(Contact("cliente@gmail.com", "Nome do Cliente"))
        sendEmailProvider.from = Contact("loja@gmail.com", "Nome do Estabelecimento")
        sendEmailProvider.dialogMessage = "Enviando comprovante"
        sendEmailProvider.connectionCallback = object : StoneCallbackInterface {
            override fun onSuccess() {
                Toast.makeText(applicationContext, "Enviado com sucesso", Toast.LENGTH_LONG).show()
            }

            override fun onError() {
                Toast.makeText(applicationContext, "Nao enviado", Toast.LENGTH_LONG).show()
            }
        }
        sendEmailProvider.execute()
    }

    private fun printReceipt(receiptType: ReceiptType, transactionObject: TransactionObject) {
        PrintController(
            applicationContext,
            PosPrintReceiptProvider(
                applicationContext,
                transactionObject, receiptType
            )
        ).print()
    }
}
